### PR TITLE
Added Testing for demo and review

### PR DESCRIPTION
**Pull request by vr-varad**
# The Pull Request is ready

- [ ] fixes #1062
- [ ] all actions are passing
- [ ] only fixes a single issue

## Framework

- [ ] the change breaks no interface
- [ ] default behaviour did not change
- [ ] any new text output is added to the translation files (at least the english one)
- [x] tests have been added (if required)
- [ ] documentation has been adjusted (if required)
- [ ] shared code has been extracted in a different file

### CODE REVIEW

**Comment by vr-varad** [comment]
@Idrinth , I've prepared initial test cases for the afterDatasetsUpdate and beforeDatasetDraw functionalities. Before proceeding with additional test cases covering edge cases, I'd appreciate your review of the existing ones.

File: test/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    expect(meta).to.have.property("$stdevFiller");
+    expect(meta.$stdevFiller).to.be.ok
+    expect(meta.$stdevFiller).to.deep.equal({
+      visible: true,
+      index: 0,
+      color: "rgba(0,0,0,0.1)",
+      above: [11, 10.1, 14, 22, 11.2],
+      below: [9, 7.9, 8, 4, 10.8],
+    });
```
[the remark this answers is not in the document]
Comment (vr-varad) [reply test/index.test.ts]: I just made a test case for review.

[the remark this answers is not in the document]
Comment (Idrinth) [reply test/index.test.ts]: don't worry too much about the rabbit @vr-varad - it is often helpful, but just an assistance :)

File: package.json [hunk 2]
```diff
@@ -19,12 +19,20 @@
   },
   "homepage": "https://idrinth-api-ben.ch",
   "scripts": {
-    "prepublishOnly": "tsc -p tsconfig.json"
+    "prepublishOnly": "tsc -p tsconfig.json",
+    "test": "mocha --timeout 30000 --require ts-node/register test/**.test.ts"
```
Comment (Idrinth) [inline package.json]: since we have a test folder, the `.test.ts` could be simplified to `.ts`

File: package.json [hunk 3]
```diff
@@ -19,12 +19,20 @@
...
   },
   "devDependencies": {
+    "@types/chai": "^4.3.16",
+    "@types/mocha": "^10.0.6",
+    "chai": "^5.1.1",
+    "mocha": "^10.4.0",
     "typescript": "^5.4.5"
+  },
+  "dependencies": {
+    "ts-node": "^10.9.2"
```
Comment (Idrinth) [inline package.json]: this is a dev dependency, only used for running tests

File: test/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import { expect } from "chai";
+import { Chart } from "chart.js";
+
+import ChartUtils from "../src/index.ts";
+
+describe("Testing ChartUtils", () => {
+  it("Testing afterDatasetsUpdate", () => {
+
+    const _meta = {
```
Comment (Idrinth) [inline test/index.test.ts]: would propose removing the underscore, this is just another const, nothing special :)

File: test/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          },
+        ],
+      },
+      getDatasetMeta: (index) => _meta,
+      isDatasetVisible: (index) => true,
+    };
+
+    ChartUtils.afterDatasetsUpdate(chart);
+
+    const meta = chart.getDatasetMeta(0);
```
Comment (Idrinth) [inline test/index.test.ts]: this call is unnecessary, `meta === _meta` due to objects being passed along by reference

File: test/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    expect(meta).to.have.property("$stdevFiller");
+    expect(meta.$stdevFiller).to.be.ok
```
Comment (Idrinth) [inline test/index.test.ts]: isn't this line covered by the next?

File: test/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      index: 0,
+      color: "rgba(0,0,0,0.1)",
+      above: [11, 10.1, 14, 22, 11.2],
+      below: [9, 7.9, 8, 4, 10.8],
+    });
+  });
+
+  it("Testing beforeDatasetDraw", () => {
+    
+    const _meta = {
```
Comment (Idrinth) [inline test/index.test.ts]: would propose removing the underscore, this is just another const, nothing special :)

**Review by Idrinth — CHANGES_REQUESTED** [review]
Very good start!

File: test/index.test.ts (hunk 6 above)
Comment (vr-varad) [reply test/index.test.ts]: actually first checks that is meta have a property and second one checks if meta.@stdevFiller have some value

File: test/index.test.ts (hunk 7 above)
Comment (vr-varad) [reply test/index.test.ts]: yeah sure

File: test/index.test.ts (hunk 6 above)
Comment (Idrinth) [reply test/index.test.ts]: I was referring to 34 and 35, the first is included in the second as far as I know.

File: test/index.ts [hunk 8]
```diff
@@ -1,12 +1,11 @@
...
-import { Chart } from "chart.js";
-
 import ChartUtils from "../src/index.ts";
+import sinon from "sinon";
 
 describe("Testing ChartUtils", () => {
   it("Testing afterDatasetsUpdate", () => {
 
-    const _meta = {
+    let meta = {
```
Comment (Idrinth) [inline test/index.ts]: this can stay a const, you are not changing the value(i.e. pointer), you are changing values within.

File: test/index.ts [hunk 9]
```diff
@@ -70,13 +66,18 @@ describe("Testing ChartUtils", () => {
         x: { getPixelForValue: (value) => value * 10 },
         y: { getPixelForValue: (value) => 100 - value * 10 },
       },
-      getDatasetMeta: (index) => _meta,
+      getDatasetMeta: (index) => meta,
     };
 
-    const args = { meta: _meta };
+    const args = { meta: meta };
```
Comment (Idrinth) [inline test/index.ts]: you can shorten the right hand side to `{ meta }`

**Review by Idrinth — CHANGES_REQUESTED** [review]

**Comment by vr-varad** [comment]
Done @Idrinth

File: test/index.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    const args = { meta };
+
+    ChartUtils.beforeDatasetDraw(chart, args);
+
+    expect(mockCtx.fillStyle).to.equal("transparent");
+    expect(mockCtx.beginPath.calledOnce).to.be.true;
+    expect(mockCtx.moveTo.calledOnce).to.be.true;
+    expect(mockCtx.lineTo.called).to.be.true;
+    expect(mockCtx.closePath.calledOnce).to.be.false;
+    expect(mockCtx.fill.calledOnce).to.be.true;
```
[the remark this answers is not in the document]
Comment (vr-varad) [reply test/index.ts]: actually with respect to beforeDatasetDraw it has not been called so testing goes negative if I make it true.

[the remark this answers is not in the document]
Comment (vr-varad) [reply test/index.ts:81]: It is giving false for called also.
